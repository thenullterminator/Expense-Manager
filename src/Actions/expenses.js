import uuid from 'uuid';

// Action Generators for expense reducers.
const addExpense=({description='',note='',amount=0,createdAt=0}={})=>{

      return {
                   type:'ADD_EXPENSE',
                   expense: {
                        id:uuid(),
                        description,
                        note,
                        amount,
                        createdAt
             }
       };
 };
 
 const removeExpense=({id}={})=>{
 
       return{
                   type:'REMOVE_EXPENSE',
                   id
        };
 };
 
 const editExpense=(id,updates)=>{
       return{
             type:'EDIT_EXPENSE',
             id,
             updates
       };
 
 };

export {addExpense,removeExpense,editExpense};