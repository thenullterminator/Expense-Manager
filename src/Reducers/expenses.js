// Expense Reducer
const expenseDefaultState=[];
const expenseReducer=(state=expenseDefaultState,action)=>{

      switch(action.type)
      {
            case 'ADD_EXPENSE':
                  return [...state,action.expense];
            case 'REMOVE_EXPENSE':
                  return state.filter(({id})=>id!==action.id);
            case 'EDIT_EXPENSE' :
                  return state.map((expense)=>{
                        if(expense.id===action.id)
                        {
                              return{
                                    ...expense,
                                    ...action.updates
                              };
                        }
                        else
                        {
                              return expense;
                        }
                  });
            default:
                  return state;
      }

};

export default expenseReducer;