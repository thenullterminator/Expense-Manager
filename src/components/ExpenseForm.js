import React from 'react';
import moment from 'moment';
import ModernDatepicker from 'react-modern-datepicker';

console.log(moment().valueOf());
export default class ExpenseForm extends React.Component{

      state={
            description:'',
            note:'',
            amount:'',
            createdAt:moment(),
            error:''
      };

      onDescriptionChange=(e)=>{
            this.setState({
                  description:e.target.value
            });
      };

      onNoteChange=(e)=>{
            this.setState({
                  note:e.target.value
            });
      };

      onAmountChange=(e)=>{

            const amount=e.target.value;
            if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
            {
                  this.setState({
                        amount
                  });
            }
            
      };

      onDateChange=(date)=>{
            if(date)
            {
                  this.setState({createdAt:date});     
            }
      };

      onSubmitForm=(e)=>{

            e.preventDefault();
            if(!this.state.description || !this.state.amount)
            {
                  this.setState({error:'Please provide both Description and Amount.'});
            }
            else
            {
                  this.setState({error:''});

                  this.props.onSubmit({
                        description:this.state.description,
                        amount:parseFloat(this.state.amount)*100,
                        createdAt:this.state.createdAt.valueOf(),
                        note:this.state.note
                  });
            }
      };

      render(){
            return (
                  <div>
                        {this.state.error && <p>{this.state.error}</p>}
                        <form onSubmit={this.onSubmitForm}>
                              <input
                                    type='text'
                                    placeholder='Description'
                                    autoFocus
                                    value={this.state.description}
                                    onChange={this.onDescriptionChange}
                              />
                              <input
                                    type='text'
                                    placeholder='Amount'
                                    value={this.state.amount}
                                    onChange={this.onAmountChange}
                              />
                              <p></p>
                              <textarea 
                                    placeholder='Add a note for the expense'
                                    value={this.state.note}
                                    onChange={this.onNoteChange}
                              ></textarea>

                              <ModernDatepicker
                                    date={this.state.createdAt.format('DD-MM-YYYY')}
                                    format={'DD-MM-YYYY'}
                                    showBorder
                                    onChange={this.onDateChange}
                                    placeholder={'Select a date'}
                              />
                              <button>Add Expense</button>
                        </form>
                  </div>
            );
      };
}