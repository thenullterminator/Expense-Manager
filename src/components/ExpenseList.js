import React from  'react';
import {connect} from 'react-redux';

import selectExpenses from '../selectors/selectors';
import ExpenseListItem from './ExpenseListItem'

const ExpenseList=(props)=>(

      <div>
            <p>This is expense list</p>
            {props.expenses.map((expense,index)=>(<ExpenseListItem key={index} {...expense} />))}
      </div>

);

const mapStateToProps=(state)=>{

      return {
            expenses:selectExpenses(state.expenses,state.filters)
      };
};

export default connect(mapStateToProps)(ExpenseList);