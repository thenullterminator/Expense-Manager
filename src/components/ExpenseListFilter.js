import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../Actions/filters';
import 'react-dates/initialize';
import {DateRangePicker} from 'react-dates';

class ExpenseListFilter extends React.Component{

      state={
            calenderFocussed:false
      };
      onDateChange=({ startDate, endDate })=>{
            if(startDate){
                  this.props.setStartDate(startDate);
            }
            if(endDate){
                  this.props.setEndDate(endDate);
            }
      };
      onFocusChange=(focusedInput)=>{
            this.setState({calenderFocussed:focusedInput});
      };
      onTextFilterChange=(e)=>{
            this.props.setTextFilter(e.target.value);
      };
      onSortByChange=(e)=>{

            if(e.target.value==='date'){
                  this.props.sortByDate();
            }else{
                  this.props.sortByAmount();
            }
      }
      render(){
            return (
                  <div>
                        <input 
                              type='text' 
                              value={this.props.filters.text} 
                              onChange={this.onTextFilterChange}
                        />

                        <select 
                        value={this.props.filters.sortBy} 
                        onChange={this.onSortByChange}
                        >
                              <option value='date'>Date</option>
                              <option value='amount'>Amount</option>
                        </select>

                        <DateRangePicker
                        startDate={this.props.filters.startDate} 
                        endDate={this.props.filters.endDate} 
                        onDatesChange={this.onDateChange} 
                        focusedInput={this.state.calenderFocussed} 
                        onFocusChange={this.onFocusChange} 
                        showClearDates={true}
                        showDefaultInputIcon={true}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                        />
                  </div>
            );
      }
}

const mapStateToProps=(state)=>({
      filters:state.filters
});
const mapDispatchToProps=(dispatch)=>({
      setStartDate:(data)=>dispatch(setStartDate(data)),
      setEndDate:(data)=>dispatch(setEndDate(data)),
      setTextFilter:(data)=>dispatch(setTextFilter(data)),
      sortByDate:()=>dispatch(sortByDate()),
      sortByAmount:()=>dispatch(sortByAmount())
});

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter);