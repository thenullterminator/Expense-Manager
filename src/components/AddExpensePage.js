import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import {addExpense} from '../Actions/expenses';


 export class AddExpensePage extends React.Component{

      onSubmit=(expense)=>{
            this.props.AddExpense(expense);
            this.props.history.push('/');//Redirecting to home page.
      };

      render(){
            return (

            <div>
                  <h2>Add Expense</h2>
                  <ExpenseForm onSubmit={this.onSubmit}/>
            </div>

            );
      }
}


const mapDispatchToProps=(dispatch)=>{
      return {
            AddExpense:(expense)=>dispatch(addExpense(expense))
      };
};

export default connect(undefined,mapDispatchToProps)(AddExpensePage);
