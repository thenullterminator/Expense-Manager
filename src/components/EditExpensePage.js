import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../Actions/expenses';

export class EditExpensePage extends React.Component{

      onSubmit=(expense)=>{
            this.props.editExpense(this.props.expense.id,expense);
            this.props.history.push('/');//Redirecting to home page.
      };

      onRemove=(e)=>{
            this.props.removeExpense(this.props.expense.id);
            this.props.history.push('/');//Redirecting to home page.
      };

      render(){
            return (
                  <div>
                        <ExpenseForm
                              expense={this.props.expense}
                              onSubmit={this.onSubmit}
                        />

                        <button onClick={this.onRemove}>Remove</button>
                  </div>
            );
      }
};

const mapStateToProps=(state,props)=>{
      return{
            expense:state.expenses.filter((expense)=> (expense.id===props.match.params.id))[0]
      };
};

const mapDispatchToProps=(dispatch,props)=>({
      editExpense:(id,expense) => dispatch(editExpense(id,expense)),
      removeExpense:(id) => dispatch(removeExpense({id}))
});

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);
