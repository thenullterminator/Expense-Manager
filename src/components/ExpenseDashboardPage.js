import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummary from '../components/ExpenseSummary';
const ExpenseDashboardPage=()=>(
      <div>
            <p>This is expense dashboard page.</p>
            <ExpenseSummary/>
            <ExpenseListFilter/>
            <ExpenseList/>
      </div>
);

export default ExpenseDashboardPage;
