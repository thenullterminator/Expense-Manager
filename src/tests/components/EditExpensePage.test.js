import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses';
import {EditExpensePage} from '../../components/EditExpensePage';

let wrapper,history,editExpense,removeExpense;

beforeEach(()=>{

      editExpense=jest.fn();
      removeExpense=jest.fn();
      history={push:jest.fn()};
      wrapper=shallow(<EditExpensePage editExpense={editExpense} removeExpense={removeExpense} history={history} expense={expenses[0]} />);
});

test('should render edit Expense Page correctly',()=>{

      expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense correctly',()=>{

      wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
      expect(history.push).toHaveBeenLastCalledWith('/');
      expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id,expenses[0]);
});

test('should handle remove expense correctly',()=>{

      wrapper.find('button').simulate('Click');
      expect(history.push).toHaveBeenLastCalledWith('/');
      expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
});