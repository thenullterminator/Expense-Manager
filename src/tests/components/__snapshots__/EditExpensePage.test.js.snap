// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit Expense Page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 12,
        "createdAt": -345600000,
        "description": "Rent 1",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
